README.md

```diff
@@ -629,6 +629,151 @@ Can you try updating the rendering function to also show `entry.content` and
+      `useState` hook
+
+### Step 8 completion check
+
+After this step is completed, we should be able to type a valid username like
+`xavierleroy` and see their most recent activity on GitHub appearing in the
+feed.
+
+Don't worry about data validation or error handling for now, we will fix that in
+the next step.
+
+## Step 9: Validating data with abstract modules, handling edge cases and errors
+
+### 9.1: Validate usernames
+
+Right now, we can enter any kind of string in our input, but GitHub only allows
+usernames with specific rules:
+- Max length should be 39 characters
+- Can only contain alphanumeric characters and dashes `-`
+- Can't start or finish with dash `-`
+
+Let's add some logic to validate the username. As the `App.re` module is getting
+large already, let's create a new module `Username.re`. In this module, we will
+define the following:
+
+- `type t = string`
+- A function `make` that takes a `string` and returns a `result` value:
+```reason
+let make = (username: string) => {
+  let re = [%re "/^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,37}[a-zA-Z0-9])?$/"];
+  Js.Re.test(~str=username, re) ? Ok(username) : Error();
+};
+```
+- A `toString` function that converts `Username.t` back to a string.
+
+To make sure we can't use any string as validated username incorrectly, we will
+make the type `t`
+[abstract](https://courses.cs.cornell.edu/cs3110/2021sp/textbook/modules/abstract_types.html),
+In order to do so, we will add an [interface
+file](https://reasonml.github.io/docs/en/module#interface-files). These files
+define how the module types will be visible from the outside.
+
+So let's create a new file `Username.rei` with this content:
+
+```reasonml
+type t;
+
+let make: string => result(t, unit);
+
+let toString: t => string;
+```
+Note how we are leaving `t` without a definition, this makes the type abstract.
+This way, every time we need to use a GitHub username in our app, we can use
+`Username.t` as a type, and this will guarantee that the string has gone through
+the validation before being used in the function.
+
+To use the new type, we will modify the existing `App` component:
+- Add a `validatedUsername` value that can be derived from existing `username`:
+```reasonml
+    let validatedUsername =
+      React.useMemo1(() => Username.make(username), [|username|]);
+```
+- Make sure that `fetchFeed` takes a value `Username.t` rather than a plain
+  string
+- Adapt the effect to depend on `validatedUsername`, not `username`
+- Modify the render function to pattern match on both `validatedUsername` and
+  `data` using a single `switch` statement.
+
+### 9.2: Handle edge cases and errors
+
+To improve the user experience, we'll add messages for the following cases:
+- when there are errors returned by the server
+- or when a valid user has an empty feed
+
+For this, we will have to modify `App` component in a couple of places.
+
+The `fetchFeed` function should check for the response status using
```

Comment:
I'm a little ashamed to admit I struggled with this part. I had the general idea of what to do, but it took a while for me to figure out that I needed to add `resolve` and `ignore` calls. Maybe it'll be less of an issue for participants since they won't be used to having `Promise.each`?

Reply:
I added some snippet to showcase how this modification should happen in https://github.com/ahrefs/react-alicante-workshop/pull/6/commits/1e63afcbe33c61d6e38e696b507d10400a80f14b.